src/interface/state_descriptor.cpp

```diff
@@ -238,40 +237,57 @@ bool StateDescriptor::AddSwarmValue(const std::string &value_name,
   return true;
 }
 
-bool StateDescriptor::AddField(const std::string &field_name, const Metadata &m_in) {
+bool StateDescriptor::AddFieldImpl(const VarID &vid, const Metadata &m_in) {
   Metadata m = m_in; // Force const correctness
-  if (m.IsSet(Metadata::Sparse)) {
-    int sparse_id = m.GetSparseId();
-    if (sparseMetadataMap_.count(field_name) > 0) {
-      if (sparseMetadataMap_.at(field_name).count(sparse_id) > 0) {
-        return false;
-      }
-    }
-    sparseMetadataMap_[field_name][sparse_id] = m;
+
+  const std::string &assoc = m.getAssociated();
+  if (m.getAssociated() == "") {
+    m.Associate(vid.label());
+  }
+  if (FieldPresent(vid.label()) || SparseBaseNamePresent(vid.label())) {
+    return false; // this field has already been added
   } else {
-    const std::string &assoc = m.getAssociated();
-    if (!assoc.length()) m.Associate(field_name);
-    if (metadataMap_.count(field_name) > 0) {
-      return false; // this field has already been added
-    } else {
-      metadataMap_[field_name] = m;
-    }
+    metadataMap_.insert({vid, m});
   }
+
   return true;
 }
 
-bool StateDescriptor::FlagsPresent(std::vector<MetadataFlag> const &flags,
-                                   bool matchAny) {
-  for (auto &pair : metadataMap_) {
-    auto &metadata = pair.second;
-    if (metadata.FlagsSet(flags, matchAny)) return true;
+bool StateDescriptor::AddSparsePoolImpl(const SparsePool &pool) {
+  if (pool.pool().size() == 0) {
+    return false;
+  }
```

Comment:
Yes, it's simply to prevent empty pools from being added. Since a sparse pool becomes immutable once it's added to the state descriptor, there's no point in adding an empty pool (it would not add any fields, but would prevent other variables or pools with the same base name from being added).

It was requested that `SparsePool` provides an interface where sparse ids can be added one by one. So that means it must be possible to create an empty pool (and later add sparse ids), like this:
```c++
auto pool = SparsePool("base_name", shared_meta_data);
pool.Add(3);
pool.Add(7);
// etc...
```